### ISSUE TITLE
Ability to add generators for generic types

### ISSUE THREAD

**Issue by AlexeyRaga** [hedgehogqa/fsharp-hedgehog-experimental#77]
It'd be nice if it was possible to register generators for generic types to `AutoGenConfig`.

For example, if I have my own generic type, like `Id<T>` then I can create `Gen<Id<T>>` for it, but there is currently no way to add it to `AutoGenConfig`.

Do you think something like that can be implemented?

**Comment by TysonMN** [comment]
Not sure. 

For how many values of `T` are you currently expressing that mapping efficiently?

**Comment by AlexeyRaga** [comment]
"how many" - does it really matter? 

There are many contexts, with multiple possible values...
`Id<T>`, `ApiResponse<T>`, `Tree<T>`, `ParserResult<T>`...

These can be types of properties in other data types:

```fsharp
type Payment = {
    id : Id<Payment>
    orderId : Id<Order>
    ....
}
```

Writing _specific_ generators every time is not very convenient.

And it is even less convenient when we put C# into the picture, because C# doesn't have built-in (with special support by HH) types for `Option` or `Result` and people use library types (not specially supported by HH).

Being able to register generators for generic types would trivialise and solve this problem: once defined generator would "just work" everywhere.

`FSCheck` also has this ability: they reflect generators from static properties of a class, and those could be of generic types, something like

```fsharp
type MyGenerators =
  static member IdGen<T>() : Arbitrary<Id<T>> = ...
  static member EdgeGen<TNode, TWeight>() : Arbitrary<Edge<TNode, TWeight>> = ...
```

**Comment by AlexeyRaga** [comment]
I dug up an example of how it is done with FSCheck in C#, to illustrate the value from the real life code:

```csharp
public static class PreludeArb
{
    public static Arbitrary<Either<TLeft, TRight>> EitherArb<TLeft, TRight>() =>
        Gen.OneOf(
            ArbMap.Default.GeneratorFor<TLeft>().Select(Either.Left<TLeft, TRight>),
            ArbMap.Default.GeneratorFor<TRight>().Select(Either.Right<TLeft, TRight>)).ToArbitrary();

    public static Arbitrary<Option<T>> OptionArb<T>() =>
        Gen.OneOf(
            ArbMap.Default.GeneratorFor<T>().Select(Option.Some), 
            Gen.Constant(Option.None<T>())).ToArbitrary();

    public static Arbitrary<NonEmptyList<T>> NonEmptyListArb<T>() =>
        ArbMap.Default
            .GeneratorFor<T>()
            .NonEmptyListOf()
            .Select(xs => new NonEmptyList<T>(xs[0], xs.Skip(1)))
            .ToArbitrary();
}
```

Unfortunately, doing it with Hedgehog is not really possible at the moment...

**Comment by TysonMN** [comment]
> "how many" - does it really matter?

Yes. If you want this feature so you can write one line of code instead of two, then your use case is not sufficient justification for this feature.  If you would avoid writing 100 lines of code, then that is sufficient justification.

----

Want to submit PR with this feature?

**Comment by AlexeyRaga** [comment]
> If you want this feature so you can write one line of code instead of two, then your use case is not sufficient justification for this feature. If you would avoid writing 100 lines of code, then that is sufficient justification.

It is more on a conceptual level here, as you probably see now. Can be 1 line, can be hundreds, depends on the case.
In my specific use cases it'd be a lot of repetitive lines across multiple repositories.
But, really, why `List<T>` is OK and `Id<T>` or `NonEmptyList<T>` or `Tree<T>` is suddenly not...

> Want to submit PR with this feature?

I could try, if there is a guidance on how it can/should be done in this library.

I am a bit of afraid of just starting to work on it, based on my previous experiences, where I was discouraged with some bureaucracy like splitting/joining commits, nuances about how things should/should not be written, etc.

**Comment by AlexeyRaga** [comment]
@TysonMN any chance for the PR to be merged?

**Comment by TysonMN** [comment]
@dharmaturtle and I will review.  I left a comment.
